### ISSUE TITLE
[Feature Request] Highlight intensity

### ISSUE THREAD

**Issue by vatterspun** [searchmarkers/mark-my-search#65]
The program is coming along really well and is way above what I expected.  You've already included Regex and Whole Words checks.  Just amazing ...

I say that both as a compliment and as a setup for the next issue because something that I'm running into is how the add-on is oddly a victim of it's own success.  

CURRENT BEHAVIOR

When I do a search with 6+ keywords that are more basic words, it'll essentially flag the whole page.  And now that I'm using the program almost all the time, highlighting the entire page has sort of the of just making it a pastel monster.  It makes it hard to read and the best option is to manually zap keyword highlights until I get down to the main bits.

I can add some screenshots of recent searches that I'm doing if needed.

REQUESTED BEHAVIOR

If there's a way to make the results either A) less intense or B) turn up the transparency, that would be ideal.  I wouldn't want to make the term listing up at the top any more transparent, just the words it's highlighting in page.  So there's just a soft highlight for the key terms.

---

Hope that's an interesting suggestion, thanks for your work on this.  More to come

**Comment by ator-dev** [comment]
Thank you for the continued interest and suggestions, it's great to know the work is appreciated!

I will say that I have been concerned about bloat recently, mainly since I have been a bit enthusiastic about adding new features. I think it's OK at the moment, but I'm being more cautious now and too much power can be a bad thing of course. That said, I like this idea - I'm always for more personalisation, and if this would definitely be useful.

Here are my thoughts (assuming that I implement a direct fix):

- Highlight colours were optimised to work for dark- and light-themes _without having to detect the theme and adapt_.
  - Side effect that they can be slightly too transparent on white backgrounds, and too bright on dark backgrounds
- **Solution:** toggle button/shortcut to make highlights dimmer.
  - Good for control since user knows best when brighter/dimmer is better
  - Steeper learning curve, pages with light _and_ dark backgrounds do not work well
- **Solution:** global setting to make highlights dimmer.
  - Good for users with specific visual impairments or who use MMS for complex research
  - Many users will never find the setting, only optimised for a single use case at once
- **Solution:** make highlights dimmer when there are many keywords.
  - No interaction required (automatic)
  - May confuse users, works poorly for large numbers of sparse keywords
- **Solution:** make highlights dimmer where they appear in groups.
  - Expands on an existing MMS feature ('blocks' of text are interpreted as a single flow), automatic, completely removes problem in _most_ cases
  - Maybe not actually what you want?

**Comment by ator-dev** [comment]
By the way, I'm not sure if you know this, but there are other possible workflows in Mark My Search. You can hide/show highlights without removing them using Alt+Shift+D, and you can set **hidden** to be the default:

![image](https://user-images.githubusercontent.com/66553618/203348109-565f8a85-0e7b-4375-999f-14011945b7a4.png)

![image](https://user-images.githubusercontent.com/66553618/203348240-dbccb7e6-9241-4e53-ab77-8b3dcfac6b02.png)

With highlights hidden, there are still two built-in tools to let you find matches:

* The **scroll markers** beside the scrollbar show where groups of highlights occur. 
![image](https://user-images.githubusercontent.com/66553618/203348753-69c1c3e8-bb80-4d85-8f93-c0faebcedc40.png)
* Pressing a keyword in the toolbar focuses its next occurrence in the page, starting from the last place you clicked/selected. [Alt+Shift+1..9,0] does the same thing for individual keywords, and using Ctrl instead of Alt takes you to the previous one. Shift+Space / Alt+Shift+Space is again the same but for _any_ keyword.

You can use these instead of looking for highlights manually. When 'jumping' using the buttons, highlights are temporarily shown for the group which was focused.

**Comment by vatterspun** [comment]
Oof, yeah your possible solutions are a spot-on analysis of possible paths and problems around each choice.  Very impressive.  
 
I personally want to have have something in the interface that would let me modify things on the fly.  Something like this:

![image](https://user-images.githubusercontent.com/1289134/206936410-b33f8cd0-a9fe-4790-8f22-f5f70b1a391d.png)

But I can see why that wouldn't be ideal for the average user and might clog the interface.  We've already got a lot of buttons and browser add-ons with tons of buttons generally get a cold response.

I guess the best course of action is therefore the second option of a Global Setting, but you're right that few will ever see it.  That likely makes this a low priority enhancement.

**Comment by vatterspun** [comment]
> By the way, I'm not sure if you know this, but there are other possible workflows in Mark My Search.

No I had no idea.  Again, very impressive.  Thank you.

**Comment by ator-dev** [comment]
> I personally want to have have something in the interface that would let me modify things on the fly

Makes sense, it would definitely fit in - I would probably put it on the left next to highlight visibility (maybe it could be one button which hides highlights when you dim them twice..?).

All buttons can be toggled in the settings, and this would be a default-off (like the globe [search] now is) - I think you're right it might clog the interface, and I'll experiment of course when I get round to it.

Button settings as of 1.10:
![image](https://user-images.githubusercontent.com/66553618/206937105-a362104a-6241-43ab-9513-6653cb4b0646.png)

**Comment by ator-dev** [comment]
Current plan as partially outlined in this issue:

__Toolbar__

- Contrast button
  - Left click to set intensity higher (1 band)
  - Right click to set intensity lower (1 band)
  - Pull down to show intensity bands, release pointer to select currently hovered

__Settings__

- Contrast button
  - Show contrast button, default off
  - Swap left/right mouse clicks on button
  - List of intensity bands in numeric form, e.g. "10,14,16"
- Default intensity band, e.g. 0
- Add large border to highlights (layout disrupting)

Note that while this is going to be a lot of additional complexity, I'll try to keep it as isolated as possible. I would consign this to a plugin for MMS, but seeing as it's an accessibility issue I'm bumping it temporarily to mainline extension work.

In the future there will probably an 'advanced toolbar options for MMS' add-on; when this is achieved, one of the first control/settings combos to be transferred in will be this contrast button. There will be a significant migration period wherein users are given plenty of time to install this add-on 'plugin' and set it up to their needs, which will be halted if user feedback is negative.
